GAPFILLING A GENOME SCALE METABOLIC MODEL USING DNNGIOR

Short tutorial (adapted from Boer et al. 2024, see below) by Maria Carolina Sisco. 

Improving genome-scale metabolic models of incomplete genomes with deep learning
Boer et al. 2024. DOI: 10.1016/j.isci.2024.111349

DNNgior: Deep Neural Network Guided Imputation of Reactomes
GSMM: Genome Scale Metabolic Model 
DNNgior uses AI to improve gap-filling by learning from the presence and absence of metabolic reactions across diverse bacterial genomes

INSTALLATION 
First, we need a Gurobi license, a linear programming solver.
1.Register at: https://www.gurobi.com/downloads/gurobi-software/
2.Get a free academic named-user-license here: https://www.gurobi.com/features/academic-named-user-license/  *you will need an institutional email*

In [ ]:
#!pip install -i https://pypi.gurobi.com gurobipy

In [ ]:
#!pip install dnngior

GAPFILL USING A COMPLETE MEDIUM
In this exercise we will gapfill (adding missing reactions) a GSMM of Blautia, a genus of anaerobic bacteria with probiotic characteristics.

Let's explore the GSMM with some basic Cobrapy commands!

In [3]:
import cobra
from cobra.io import read_sbml_model
draft_reconstruction = read_sbml_model('bh_ungapfilled_model.sbml')

'' is not a valid SBML 'SId'.


Set parameter Username
Academic license - for non-commercial use only - expires 2026-03-17


In [5]:
draft_reconstruction.summary()

Notice the objective function (biomass=growth rate). Also, there are no fluxes.

In [6]:
draft_reconstruction.optimize()

<Solution 0.000 at 0x75363878d210>

In [8]:
draft_reconstruction.medium

{'EX_cpd00067_e0': 1000.0,
 'EX_cpd00182_e0': 1000.0,
 'EX_cpd00106_e0': 1000.0,
 'EX_cpd00075_e0': 1000.0,
 'EX_cpd00064_e0': 1000.0,
 'EX_cpd00108_e0': 1000.0,
 'EX_cpd00036_e0': 1000.0,
 'EX_cpd11576_e0': 1000.0,
 'EX_cpd04097_e0': 1000.0,
 'EX_cpd00033_e0': 1000.0,
 'EX_cpd00412_e0': 1000.0,
 'EX_cpd00012_e0': 1000.0,
 'EX_cpd00118_e0': 1000.0,
 'EX_cpd01030_e0': 1000.0,
 'EX_cpd00161_e0': 1000.0,
 'EX_cpd00129_e0': 1000.0,
 'EX_cpd00971_e0': 1000.0,
 'EX_cpd01012_e0': 1000.0,
 'EX_cpd00635_e0': 1000.0,
 'EX_cpd00039_e0': 1000.0,
 'EX_cpd01155_e0': 1000.0,
 'EX_cpd00246_e0': 1000.0,
 'EX_cpd00020_e0': 1000.0,
 'EX_cpd00276_e0': 1000.0,
 'EX_cpd00082_e0': 1000.0,
 'EX_cpd00080_e0': 1000.0,
 'EX_cpd00023_e0': 1000.0,
 'EX_cpd15560_e0': 1000.0,
 'EX_cpd15561_e0': 1000.0,
 'EX_cpd00367_e0': 1000.0,
 'EX_cpd00063_e0': 1000.0,
 'EX_cpd00058_e0': 1000.0,
 'EX_cpd00107_e0': 1000.0,
 'EX_cpd00573_e0': 1000.0,
 'EX_cpd00184_e0': 1000.0,
 'EX_cpd00224_e0': 1000.0,
 'EX_cpd00314_e0': 1000.0,
 

The exchange reactions (uptake reactions) are set to a unlimited value, there is no constraint regarding what the bacteria can take from the medium, but even so, our model can't simulate growth, why?

Let's start gapfilling our model
Import the dnngior library and use the Gapfill class to gapfill the reconstruction

In [9]:
import os, sys
path_to_blautia_model = ("bh_ungapfilled_model.sbml")

In [10]:
import dnngior
gapfilled_model_complete = dnngior.Gapfill(draftModel = path_to_blautia_model, 
                                          medium = None, 
                                          objectiveName = 'bio1')

E0000 00:00:1742999174.229698    7129 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1742999174.276271    7129 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1742999174.531863    7129 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1742999174.531892    7129 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1742999174.531895    7129 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1742999174.531896    7129 computation_placer.cc:177] computation placer already registered. Please check linka

The rest of dnngior features can be used without it.


'' is not a valid SBML 'SId'.


Gap-filling database =  ModelSEED


No objective coefficients in model. Unclear what should be optimized


#reactions not found in NN-keys:  155 / 1656
Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored
Flux through biomass reaction is 1.00000000
Flux through biomass reaction is 1.00000000
Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored


 condition is currently:  77488 


Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored


 condition is currently:  38744 


Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored


 condition is currently:  19372 


Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored


 condition is currently:  9686 


Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored


 condition is currently:  4843 


Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored


 condition is currently:  2421 


Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored


 condition is currentl

Make a new object of the gapfilled model

In [11]:
gf_model_compl_med = gapfilled_model_complete.gapfilledModel.copy()

Read LP format model from file /tmp/tmpv6wwblry.lp
Reading time = 0.01 seconds
: 1401 rows, 3344 columns, 14428 nonzeros


In [12]:
gf_model_compl_med.optimize()

<Solution 146.138 at 0x75363878dab0>

#It's growing!, the growth rate after optimization is : 146.138 mmol gDW−1 hr−1 *Millimoles per gram dry cell weight per hour, the default flux units used in FBA

Now let's see how many and which reactions DNNgior added in order to simulate growth

In [15]:
print("Number of reactions added:", len(gapfilled_model_complete.added_reactions))
print("~~")
for reaction in gapfilled_model_complete.added_reactions:
    print(gf_model_compl_med.reactions.get_by_id(reaction).name)

Number of reactions added: 16
~~

EX_cpd15432_e0
core oligosaccharide lipid A transport via vector (periplasm to extracellular)
stearoyl-cardiolipin synthase
S-adenosyl-L-methione:demethylmenaquinone methyltransferase
-
acyl-CoA:malonyl-CoA C-acyltransferase (decarboxylating, oxoacyl-
EX_cpd15511_e0
2-Amino-4-hydroxy-6-(erythro-1,2,3-trihydroxypropyl) dihydropteridine triphosphate phosphohydrolase (alkaline optimum)
all-trans-polyprenyl diphosphate:1,4-dihydroxy-2-naphthoate polyprenyltransferase
R_R05617_c
acyl-CoA:malonyl-CoA C-acyltransferase (decarboxylating, oxoacyl- and enoyl-reducing, thioester-hydrolysing)
glycine/sarcosine/dimethylglycine N-methyltransferase
murein polymerizing transglycosylase
EX_cpd02229_e0
glycolaldehyde:NAD+ oxidoreductase


In [16]:
gf_model_compl_med.reactions.get_by_id('EX_cpd15432_e0')

<Reaction EX_cpd15432_e0 at 0x7535c6501a80>

In [17]:
gf_model_compl_med.reactions.get_by_id('EX_cpd15511_e0')

<Reaction EX_cpd15511_e0 at 0x7535ca67cf40>

GAPFILL USING A DEFINED MEDIUM

First, load the media file containing the composition of the medium

In [19]:
medium_file_path = 'Nitrogen-Nitrite_media.tsv'

In [20]:
import pandas as pd
new_medium = pd.read_csv(medium_file_path, sep="\t")
new_medium.head()

id       name  concentration  minflux  maxflux
0  cpd00027  D-Glucose          0.001     -100        5
1  cpd00075    Nitrite          0.001     -100        5
2  cpd00009  Phosphate          0.001     -100        5
3  cpd00048    Sulfate          0.001     -100        5
4  cpd00063       Ca2+          0.001     -100      100

Let's gapfill our GSMM so it can growth on this medium

In [21]:
gapfill_nitr = dnngior.Gapfill(path_to_blautia_model, medium_file = medium_file_path, objectiveName = 'bio1')

'' is not a valid SBML 'SId'.
'' is not a valid SBML 'SId'.


Gap-filling database =  ModelSEED


No objective coefficients in model. Unclear what should be optimized


Loading medium from: Nitrogen-Nitrite_media.tsv
#reactions not found in NN-keys:  7 / 1508
Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored
Flux through biomass reaction is 1.00000000
Flux through biomass reaction is 1.00000000
Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored


 condition is currently:  75420 


Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored


 condition is currently:  37710 


Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored


 condition is currently:  18855 


Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored


 condition is currently:  9427 


Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored


 condition is currently:  4713 


Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored


 condition is currently:  2356 


Warning for adding constraints: zero or small (< 1e-13) 

Again, make a new object of the gapfilled model and check if it's growing.

In [22]:
gf_model_Nit_med = gapfill_nitr.gapfilledModel.copy()
gf_model_Nit_med.optimize()

Read LP format model from file /tmp/tmp47a2aamf.lp
Reading time = 0.01 seconds
: 1423 rows, 3170 columns, 14782 nonzeros


<Solution 39.996 at 0x7535c6a913f0>

Let's see how many and which reactions DNNgior added in order to simulate growth on the nitrite media

In [23]:
print("Number of reactions added:", len(gapfill_nitr.added_reactions))
print("~~")
#for reaction in gapfill_nitr.added_reactions[:5]:
for reaction in gapfill_nitr.added_reactions:
    print(gf_model_Nit_med.reactions.get_by_id(reaction).name)

Number of reactions added: 77
~~
R_L_Glutamate_tRNA_Glu__ligase__AMP_forming_
5-Formiminotetrahydrofolate:L-glutamate N-formiminotransferase


-

acyl-CoA:malonyl-CoA C-acyltransferase (decarboxylating, oxoacyl-

Lumped Isoleucine Synthesis
L-Cysteine hydrogen-sulfide-lyase (deaminating; pyruvate-forming)
biotin-carboxyl-carrier-protein:carbon-dioxide ligase (ADP-forming)
dihydrodipicolinate reductase
DNA biosynthesis
R_Protein_synthesis_c
degradation of starch300 by phosphorylase, amylase, dextrinase, maltase (chloroplast), 1:3 glc:g1p (light-dep rxn)
thiazole phosphate synthesis (ribose 5-phosphate), yeast-specifc
ADP Glucose Pyrophosphorylase
Glutamyl-tRNA synthetase, mitochondrial

R_Nitrogenous_compound_synthesis_c
Phosphatidylethanolamine phosphatidohydrolase
Sink needed to allow 4-hydroxy-benzoate to leave system

EX_cpd00030_e0

EX_cpd15560_e0
all-trans-polyprenyl diphosphate:1,4-dihydroxy-2-naphthoate polyprenyltransferase
Nitrogenous compounds,NADPH:oxygen oxidoreductase (3-h

THE END!